examples/cpp/grpc/pingpong/pingpong.cpp

```diff
@@ -0,0 +1,273 @@
+    void PingShouldThrow(
+        bond::ext::gRPC::unary_call<
+        bond::bonded<PingRequest>,
+        bond::bonded<PingReply>> call) override
+    {
+        call.FinishWithError(Status(StatusCode::CANCELLED, "do not want to respond"));
+    }
+};
+
+class PingPongServiceImpl final : public PingPong<PingRequest>::Service
+{
+    void Ping(
+        bond::ext::gRPC::unary_call<
+        bond::bonded<PingRequest>,
+        bond::bonded<PingReply>> call) override
+    {
+        PingRequest request = call.request().Deserialize();
+
+        PingReply reply;
+        reply.message = "ping " + request.name;
+
+        call.Finish(bond::bonded<PingReply>{reply}, Status::OK);
+    }
+};
+
+void printAndSet(
+    countdown_event* print_event,
+    bool* isCorrectResponse,
+    const bond::bonded<PingReply>& response,
+    const Status& status)
+{
+
+    *isCorrectResponse = false;
+
+    if(status.ok())
+    {
+        const std::string& message = response.Deserialize().message;
+
+        if (message.compare("ping pong") == 0)
+        {
+            std::cout << "Correct response: " << message;
+            *isCorrectResponse = true;
+        }
+        else
+        {
+            std::cout << "Wrong response";
+            *isCorrectResponse = false;
+        }
+    }
+
+    print_event->set();
+}
+
+int main()
+{
+    auto ioManager = std::make_shared<io_manager>();
+    auto threadPool = std::make_shared<bond::ext::gRPC::thread_pool>();
+
+    DoublePingServiceImpl double_ping_service;
+    PingPongServiceImpl ping_pong_service;
+
+    bond::ext::gRPC::server_builder builder;
+    builder.SetThreadPool(threadPool);
+    const std::string server_address("127.0.0.1:50051");
+    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
+    builder.RegisterService(&double_ping_service);
+    builder.RegisterService(&ping_pong_service);
+    std::unique_ptr<bond::ext::gRPC::server> server(builder.BuildAndStart());
+
+    DoublePing::Client doublePing(
+        grpc::CreateChannel(server_address, grpc::InsecureChannelCredentials()),
+        ioManager,
+        threadPool);
+
+    PingPong<PingRequest>::Client pingPong(
+        grpc::CreateChannel(server_address, grpc::InsecureChannelCredentials()),
+        ioManager,
+        threadPool);
+
+
+    const std::string user("pong");
+
+    countdown_event ping_event(5);
+
+    bool isCorrectResponse;
+    bool pingNoPayloadIsCorrectResponse;
+    bool pingVoidIsCorrectResponse;
+    bool pingShouldThrowIsCorrectResponse;
+    bool pingGenericIsCorrectResponse;
+
+    ClientContext context;
+    ClientContext pingNoPayloadContext;
+    ClientContext pingNoResponseContext;
+    ClientContext pingVoidContext;
+    ClientContext pingShouldThrowContext;
+    ClientContext pingGenericContext;
+
+    PingRequest request;
+    request.name = user;
+    bond::bonded<PingRequest> req(request);
+
+    {
+        std::function<void(const bond::bonded<PingReply>&, const Status&)> f_print =
```

Comment:
we don't need it, but I liked how it read better separated. I will replace it to auto though